### PR TITLE
Enable structured types as config values

### PR DESCRIPTION
**Pull request by lukas-c-wilhelm**
This commit enables structured types as config values. That is, something like
```
@BaiganConfig
public interface SomeComplexConfiguration {
    SomeConfigObject someConfig();
}
```
can now be deserialized, e.g. from a config file with content
```
[{ "alias": "some.complex.configuration.some.config", "defaultValue": {"config_key":"a value"}}]
```

Users of the library do not have to add any extra configuration to enable structured config types.

It works like this:
- There is a class `BaiganConfigClasses` that provides a mapping of config keys to their types. It is provided as a bean that is constructed by the `ContextAwareConfigurationMethodInvocationHandler` when creating the proxies for the `BaiganConfig` interfaces.
- The repository is no longer handling the deserialization. Instead, it only reads the data from the source and hands it to a bean of type `ConfigurationParser`.
- The `ConfigurationParser` does the actual deserialization of the config values. It first deserializes the data to `Configuration<JsonNode>`, then deserializes the `JsonNode` config value to the correct class for each config key, using a bean of type `BaiganConfigClasses` to determine the right type. 
- The `ConfigurationParser` is passed to the repositories by the `RepositoryFactory`. Library users do not need to be aware of these details.
- The repository builders accept an `ObjectMapper` instance that will be used for the deserialization of the configuration.

The change is backwards compatible with respect to config files. All old config files can be processes as before. This is ensured in End2End tests.

TODOs
- [x] Check test coverage for changes
- [x] End2End tests for other repositories
- [x] Update documentation

### CODE REVIEW

**Comment by lukas-c-wilhelm** [comment]
@lukasniemeier-zalando do we want to add a component test for the `GitConfigurationRepo` or just remove it since it is marked deprecated anyway?

File: src/main/java/org/zalando/baigan/service/ConfigurationParser.java
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/baigan/service/ConfigurationParser.java:31]: Can we go one step further and have the whole `BaiganConfigClasses` mechanism hidden from library users? As far as I can tell in `ConfigurationBeanDefinitionRegistrar` the library is providing and configuring the respective (singleton) bean for the application already. What if we, for example, let the `ConfigurationParser` obtain the bean from the application context for example?

File: src/main/java/org/zalando/baigan/service/FileSystemConfigurationRepository.java
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/baigan/service/FileSystemConfigurationRepository.java:43]: At this point it probably makes sense to provide a builder interface equally how we provide on for the S3 repository, wdyt?

Comment (lukasniemeier-zalando) [reply src/main/java/org/zalando/baigan/service/FileSystemConfigurationRepository.java:43]: (1) may be a separate (maybe earlier PR) (2) see also my comment regarding "who is providing the `BaiganConfigClasses` bean"

**Review by lukasniemeier-zalando — COMMENTED** [review]
I like the change as it moves deserialization to the pre-caching layer, which I believe is a crucial change once we allow complex type (deserialization). It's a considerably large change that I expect us to iterate a bit on, we may have to communicate this to library users (and find an early adopter application for validation).

I am interested in the thoughts of other baigan maintainers (if there are any left :cry:) regarding if one _want_ to support complex types with baigan. I know that the original intention was to only support simple feature flags _by design_ (`true`/`false` or `0`/`100`).

**Comment by lukasniemeier-zalando** [comment]
> do we want to add a component test for the GitConfigurationRepo or just remove it since it is marked deprecated anyway?

My vote goes for removing it (in a separate, maybe earlier) PR. We have discourage from usage for many years

**Comment by lukas-c-wilhelm** [comment]
> > do we want to add a component test for the GitConfigurationRepo or just remove it since it is marked deprecated anyway?
> 
> My vote goes for removing it (in a separate, maybe earlier) PR. We have discourage from usage for many years

Done in https://github.com/zalando-stups/baigan-config/pull/78

File: src/main/java/org/zalando/baigan/service/ConfigurationParser.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/service/ConfigurationParser.java:31]: Now that we have the ConfigurationParser, this should indeed be quite conveniently possible. I'll have a look into this.

**Comment by lukas-c-wilhelm** [comment]
> I like the change as it moves deserialization to the pre-caching layer, which I believe is a crucial change once we allow complex type (deserialization). It's a considerably large change that I expect us to iterate a bit on, we may have to communicate this to library users (and find an early adopter application for validation).
> 
> I am interested in the thoughts of other baigan maintainers (if there are any left 😢) regarding if one _want_ to support complex types with baigan. I know that the original intention was to only support simple feature flags _by design_ (`true`/`false` or `0`/`100`).

Good call out. I'd appreciate input by other maintainers as well. The reasoning is that I see Baigan applied not just as a feature flag library, but as a general library for business configuration, where each file contains the configuration for a particular use case. This is a lot nicer if structured configuration types are possible.

File: src/main/java/org/zalando/baigan/service/FileSystemConfigurationRepository.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/service/FileSystemConfigurationRepository.java:43]: Addressed in https://github.com/zalando-stups/baigan-config/pull/79

File: src/main/java/org/zalando/baigan/service/EtcdConfigurationRepository.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/service/EtcdConfigurationRepository.java:32]: To be able to do this, I first need end-2-end tests. This, however, requires upgrading to etcd v3, since I failed to get a docker image running with Testcontainers for version 2. I would propose to first make a separate PR to upgrade etcd to version 3, then proceed to implementing structured config types in this PR. What do you think?

File: src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java
Comment (kasmarian) [inline src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java:99]: Speaking about breaking changes. For complex configurations in my team we were using a constructor with `LinkedHashMap<String, ?> rawData` argument and then converted the payload to what we needed. Only LinkedHashMap worked.

Example:
```
{
     "alias": "feature.config.default.data",
     "defaultValue": {
      "a8a23682-1623-450b-8817-50c98827ea4e": [
        {
          "id": "default_data",
          "components": {
            "componentA": [
              {
                "targets": ["A", "B"],
                "data": {
                  "subdata1": "something 1",
                  "subdata2": "something 2",
                  "subdata3": [
                    {
                      "key1": "value1",
                      "key2": "value2"
                    },
                    {
                      "key1": "value3",
                      "key2": "value4"
                    }
                  ],
                  "media": {
                    "mobile": {
                      "src": "https://mobile-image.jpg"
                    },
                    "desktop": {
                     "src": "https://desktop-image.jpg"
                    }
                  },
                  ... some more fields here
                }
              }
...
```

Then the baigan interface looks like
```
@BaiganConfig
public interface FeatureConfig {
    FeatureConfigData defaultData();
}
```
And the actual class with data:
```
public class FeatureConfigData {

    private static final ObjectMapper MAPPER = JacksonConfig.getObjectMapper()
            .setPropertyNamingStrategy(PropertyNamingStrategies.LOWER_CAMEL_CASE);

    private final Map<UUID, List<FeatureConfig>> data;

    public FeatureConfigData(@NonNull LinkedHashMap<String, List<Map<String, ?>>> rawData) {
        data = MAPPER.convertValue(rawData, new TypeReference<>() {
        });
    }
```

The rest of the structure of inner classes is irrelevant as Jackson will take care of them. 


With this change, it looks like, these workarounds will fail, and hopefully, they will fail on build time, as team should have tests covering this configuration. But, I think, it is still important to announce the fact of the breaking change explicit somehow. 

I check an example on this PR, and looks like with some minor adjustments it's possible to still keep custom initialization logic, if needed, if  `FeatureConfigData` constructor will be annotated with `@JsonCreator`.

File: src/main/java/org/zalando/baigan/service/ChainedConfigurationRepository.java
Comment (kasmarian) [inline src/main/java/org/zalando/baigan/service/ChainedConfigurationRepository.java:53]: A bit of irrelevant to the PR question, but what was the intention of this method? All implementations throw `UnsupportedOperationException` :thinking:

**Review by kasmarian — COMMENTED** [review]
Hi, first, I'd like to note that I love the change, and it's something that is truly useful to have supported by the library out of the box. In my team, we used complex configs for a long time, but we're applying workarounds. I've added questions related to the backward compatibility of such workaround with this feature release.

File: src/main/java/org/zalando/baigan/service/ChainedConfigurationRepository.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/service/ChainedConfigurationRepository.java:53]: I was wondering the same, but decided to ignore it for this PR. Looks like the results of a YAGNI case to me :grin:

File: src/main/java/org/zalando/baigan/proxy/ConfigurationBeanDefinitionRegistrar.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/proxy/ConfigurationBeanDefinitionRegistrar.java]: I was able to make this work using generic return type with quite little changes.

File: src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java:99]: Yes, I would think that everything should work fine if only `FeatureConfigData` can be properly deserialized by the object mapper. I also plan to make the object mapper configurable to make this more flexible.

You are right to point out that we have to highlight this when releasing, thanks a lot. We can work out what exactly to recommend when writing documentation and finalizing the change.

File: src/main/java/org/zalando/baigan/proxy/ConfigurationBeanDefinitionRegistrar.java
Comment (kasmarian) [reply src/main/java/org/zalando/baigan/proxy/ConfigurationBeanDefinitionRegistrar.java]: ah, great! I overestimated the size of the change

**Comment by lukas-c-wilhelm** [comment]
I extracted the test coverage improvements to a separate PR https://github.com/zalando-stups/baigan-config/pull/80 to further slim down this one here.

File: src/main/java/org/zalando/baigan/service/EtcdConfigurationRepository.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/service/EtcdConfigurationRepository.java:32]: I extracted the test coverage improvements to a separate PR https://github.com/zalando-stups/baigan-config/pull/80 to further slim down this one here. When we have a decision on the etcd repo, I will potentially make a separate PR with version upgrade and tests.

File: src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java:99]: `ObjectMapper' can now optionally be provided as bean `baiganObjectMapper`.

File: src/main/java/org/zalando/baigan/service/EtcdConfigurationRepository.java
Comment (lukasniemeier-zalando) [reply src/main/java/org/zalando/baigan/service/EtcdConfigurationRepository.java]: Sounds good to me - also I am not necessarily saying that we need to block this PR until we have resolved all other possible dependencies, totally fair to me if we say "this and that become TODOs" :next_track_button:

File: src/main/java/org/zalando/baigan/service/ConfigurationParser.java
Comment (lukasniemeier-zalando) [reply src/main/java/org/zalando/baigan/service/ConfigurationParser.java]: Top!

File: src/main/java/org/zalando/baigan/service/ChainedConfigurationRepository.java
Comment (lukasniemeier-zalando) [reply src/main/java/org/zalando/baigan/service/ChainedConfigurationRepository.java]: Original usages of baigan-config used the same library code to implement processes to update the configuration (repositories), i.e. the applications themselves were toggling off feature toggles when certain conditions were met.

File: src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java:99]: > I check an example on this PR, and looks like with some minor adjustments it's possible to still keep custom initialization logic, if needed, if FeatureConfigData constructor will be annotated with @JsonCreator.

I made an end-2-end test and it looks like annotating the constructor with `@JsonCreator` is actually sufficient to make it work. I will not add this test to the PR as it is somewhat artificial. Generally, any class that the object mapper can deserialize from JSON can be used as type for a configuration.

File: src/main/java/org/zalando/baigan/service/ConfigurationParser.java
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/baigan/service/ConfigurationParser.java:36]: Do we have a test (or shall we add one) that verifies that an application context offering a specifically qualified `ObjectMapper` bean is handing it to the `ConfigurationParser` (plus the negative variants of that validation)?

**Review by lukasniemeier-zalando — APPROVED** [review]
Looks good to me, many thanks! In terms of open tasks I see the "documentation one", for which I feel a light refresh of the `README` can also be a dedicated PR (we also still have https://github.com/zalando-stups/baigan-config/pull/78#issuecomment-1752482911). Besides that I guess it's just resolving the merge conflicts now, then we take it in.

File: src/main/java/org/zalando/baigan/service/ConfigurationParser.java
Comment (lukas-c-wilhelm) [reply src/main/java/org/zalando/baigan/service/ConfigurationParser.java:36]: There is a test that does it, but since it does not special configuration of the object mapper, it does not actually validate it properly. I'll extend this to cover the case.

**Comment by lukas-c-wilhelm** [comment]
> Looks good to me, many thanks! In terms of open tasks I see the "documentation one", for which I feel a light refresh of the `README` can also be a dedicated PR (we also still have [#78 (comment)](https://github.com/zalando-stups/baigan-config/pull/78#issuecomment-1752482911)). Besides that I guess it's just resolving the merge conflicts now, then we take it in.

Can you also check the change proposal https://github.com/zalando-stups/baigan-config/pull/82

The current PR makes using the `ChainedConfigurationRepository` error prone as you can construct the individual repos with the constructor, but if they are not created as beans, the `ConfigurationParser` is not actually set and they don't work. The PR linked above fixes this at the cost of exposing a `RepositoryFactory` bean.

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Pull request merged**
